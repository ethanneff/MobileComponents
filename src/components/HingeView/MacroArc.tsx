import React, { memo } from "react";
import { Circle, G, Path } from "react-native-svg";
import Theme from "../../commons/Theme";
import { polarToCartesian, wrapTo360 } from "../../commons/Utils/Trigo";

interface MacroArcProps {
  startAngle: number | undefined;
  endAngle: number | undefined;
  centerX: number | undefined;
  centerY: number | undefined;
  limbLength: number | undefined;
  scaledDown?: boolean;
}

const styles = (scaledDown?: boolean) => ({
  arc: {
    strokeColor: Theme.color.neutral50,
    strokeWidth: 40,
    radius: scaledDown ? 20 : 24
  }
});

export default memo(function MacroArc({
  startAngle,
  endAngle,
  centerX,
  centerY,
  limbLength,
  scaledDown
}: MacroArcProps) {
  if (
    startAngle === undefined ||
    endAngle === undefined ||
    centerX === undefined ||
    centerY === undefined ||
    limbLength === undefined
  ) {
    return null;
  }
  const arc = styles(scaledDown).arc;
  const arcRadius = limbLength * 2 - arc.strokeWidth / 2;

  const { arcCoordinates, arcEnd } = describeMacroArc(
    centerX,
    centerY,
    arcRadius,
    startAngle,
    endAngle
  );

  const arcEndCircleProps = {
    r: arc.strokeWidth / 2,
    stroke: arc.strokeColor,
    strokeWidth: 4,
    fill: arc.strokeColor,
    opacity: 1
  };
  return (
    <G>
      <Path
        strokeWidth={arc.strokeWidth}
        d={arcCoordinates}
        fill={"none"}
        stroke={arc.strokeColor}
      />
      <Circle
        id={"macro-arc-end"}
        cx={arcEnd.x}
        cy={arcEnd.y}
        {...arcEndCircleProps}
      />
    </G>
  );
});

function describeMacroArc(
  x: number,
  y: number,
  radius: number,
  startAngle: number,
  endAngle: number
) {
  const start = wrapTo360(startAngle);
  const end = wrapTo360(endAngle);

  const arcStart = polarToCartesian(x, y, radius, end);
  const arcEnd = polarToCartesian(x, y, radius, start);

  const largeArcFlag = "0";
  const sweepFlag = end - start < 180 ? "0" : "1";

  const arcCoordinates = [
    "M",
    arcStart.x,
    arcStart.y,
    "A",
    radius,
    radius,
    0,
    largeArcFlag,
    sweepFlag,
    arcEnd.x,
    arcEnd.y
  ].join(" ");

  return { arcCoordinates, arcStart, arcEnd };
}
