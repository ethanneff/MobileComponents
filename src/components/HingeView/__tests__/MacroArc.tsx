import { render } from "@testing-library/react-native";
import React from "react";
import MacroArc from "../MacroArc";

describe("MacroCurve", () => {
  it("renders with the correct data", () => {
    const { container } = render(
      <MacroArc
        startAngle={90}
        endAngle={0}
        centerX={800}
        centerY={500}
        limbLength={50}
        scaledDown={false}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
