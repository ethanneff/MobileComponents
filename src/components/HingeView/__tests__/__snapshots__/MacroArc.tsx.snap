// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MacroCurve renders with the correct data 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNSVGGroup
    fill={4278190080}
    fillOpacity={1}
    fillRule={1}
    font={Object {}}
    matrix={
      Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ]
    }
    opacity={1}
    originX={0}
    originY={0}
    propList={Array []}
    rotation={0}
    scaleX={1}
    scaleY={1}
    skewX={0}
    skewY={0}
    stroke={null}
    strokeDasharray={null}
    strokeDashoffset={null}
    strokeLinecap={0}
    strokeLinejoin={0}
    strokeMiterlimit={4}
    strokeOpacity={1}
    strokeWidth={1}
    vectorEffect={0}
    x={0}
    y={0}
  >
    <RNSVGPath
      d="M 880 500 A 80 80 0 0 0 800 580"
      fill={null}
      fillOpacity={1}
      fillRule={1}
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
      opacity={1}
      originX={0}
      originY={0}
      propList={
        Array [
          "fill",
          "stroke",
          "strokeWidth",
        ]
      }
      rotation={0}
      scaleX={1}
      scaleY={1}
      skewX={0}
      skewY={0}
      stroke={4293783794}
      strokeDasharray={null}
      strokeDashoffset={null}
      strokeLinecap={0}
      strokeLinejoin={0}
      strokeMiterlimit={4}
      strokeOpacity={1}
      strokeWidth={40}
      vectorEffect={0}
      x={0}
      y={0}
    />
    <RNSVGCircle
      cx={800}
      cy={580}
      fill={4293783794}
      fillOpacity={1}
      fillRule={1}
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
      name="macro-arc-end"
      opacity={1}
      originX={0}
      originY={0}
      propList={
        Array [
          "fill",
          "stroke",
          "strokeWidth",
        ]
      }
      r={20}
      rotation={0}
      scaleX={1}
      scaleY={1}
      skewX={0}
      skewY={0}
      stroke={4293783794}
      strokeDasharray={null}
      strokeDashoffset={null}
      strokeLinecap={0}
      strokeLinejoin={0}
      strokeMiterlimit={4}
      strokeOpacity={1}
      strokeWidth={4}
      vectorEffect={0}
      x={0}
      y={0}
    />
  </RNSVGGroup>
</View>
`;
